### PR TITLE
refactor: remove localhost client implementation

### PR DESCRIPTION
**Pull request by seantking**
## Description

closes: #1089

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/core/02-client/abci.go [hunk 1]
```diff
@@ -1,42 +0,0 @@
...
-			upgradedConsState := &ibctmtypes.ConsensusState{
-				Timestamp:          ctx.BlockTime(),
-				NextValidatorsHash: ctx.BlockHeader().NextValidatorsHash,
-			}
-			bz := k.MustMarshalConsensusState(upgradedConsState)
-
-			k.SetUpgradedConsensusState(ctx, plan.Height, bz)
-		}
-	}
-
```
Comment (colin-axner) [inline modules/core/02-client/abci.go:32]: this is not related to the localhost

File: modules/core/02-client/genesis.go [hunk 2]
```diff
@@ -65,7 +60,6 @@ func ExportGenesis(ctx sdk.Context, k keeper.Keeper) types.GenesisState {
 		ClientsMetadata:    clientsMetadata,
 		ClientsConsensus:   k.GetAllConsensusStates(ctx),
 		Params:             k.GetParams(ctx),
-		CreateLocalhost:    false,
```
Comment (colin-axner) [inline modules/core/02-client/genesis.go:68]: we might actually want to leave this param in. I'm not sure what the migration process is like for removing on chain parameters

**Review by colin-axner — CHANGES_REQUESTED** [review]

File: modules/core/02-client/abci.go (hunk 1 above)
Comment (seantking) [reply modules/core/02-client/abci.go:32]: https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/abci.go#L11

This looks to me like it's only being used by the localhost client? (comments also suggest as much)

Comment (seantking) [reply modules/core/02-client/abci.go:32]: https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/abci.go#L11

This looks to me like it's only being used by the localhost client? (comments also suggest as much)

File: modules/core/02-client/genesis.go (hunk 2 above)
Comment (seantking) [reply modules/core/02-client/genesis.go:68]: I've added it back in and also added it to the `AppModule` struct again (although I'm not sure if this is necessary). From what I can tell we don't actually use the field in `InitGenesis` at all.

File: modules/core/02-client/abci.go (hunk 1 above)
Comment (colin-axner) [reply modules/core/02-client/abci.go:32]: No the godoc is incorrect. This code is required to perform IBC client upgrades, see in code documentation

File: modules/core/02-client/abci.go [hunk 3]
```diff
@@ -1,42 +0,0 @@
...
-	_, found = k.GetClientState(ctx, exported.Localhost)
-	if !found {
-		return
-	}
-
-	// update the localhost client with the latest block height
-	if err := k.UpdateClient(ctx, exported.Localhost, nil); err != nil {
-		panic(err)
-	}
```
Comment (colin-axner) [inline modules/core/02-client/abci.go:41]: these lines can be deleted

File: modules/core/02-client/genesis.go (hunk 2 above)
Comment (colin-axner) [reply modules/core/02-client/genesis.go:68]: we can remove the `AppModule` field :+1:

File: modules/core/02-client/keeper/client.go [hunk 4]
```diff
@@ -98,8 +95,7 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, header exported.C
 	// Else the update was proof of misbehaviour and we must emit appropriate misbehaviour events.
 	if status := newClientState.Status(ctx, clientStore, k.cdc); status != exported.Frozen {
 		// if update is not misbehaviour then update the consensus state
-		// we don't set consensus state for localhost client
-		if header != nil && clientID != exported.Localhost {
+		if header != nil {
 			k.SetClientConsensusState(ctx, clientID, header.GetHeight(), newConsensusState)
 		} else {
 			consensusHeight = types.GetSelfHeight(ctx)
```
Comment (colin-axner) [inline modules/core/02-client/keeper/client.go:101]: this line can be removed

File: modules/core/02-client/keeper/client_test.go [hunk 5]
```diff
@@ -25,7 +24,6 @@ func (suite *KeeperTestSuite) TestCreateClient() {
 		expPass     bool
 	}{
 		{"success", ibctmtypes.NewClientState(testChainID, ibctmtypes.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, testClientHeight, commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath, false, false), true},
-		{"client type not supported", localhosttypes.NewClientState(testChainID, clienttypes.NewHeight(0, 1)), false},
```
Comment (colin-axner) [inline modules/core/02-client/keeper/client_test.go:28]: refactor test case instead of removing

File: modules/core/02-client/keeper/keeper_test.go [hunk 6]
```diff
@@ -177,11 +169,6 @@ func (suite *KeeperTestSuite) TestValidateSelfClient() {
 			ibctmtypes.NewClientState(suite.chainA.ChainID, ibctmtypes.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, testClientHeight, commitmenttypes.GetSDKSpecs(), nil, false, false),
 			true,
 		},
-		{
-			"invalid client type",
-			localhosttypes.NewClientState(suite.chainA.ChainID, testClientHeight),
-			false,
-		},
```
Comment (colin-axner) [inline modules/core/02-client/keeper/keeper_test.go:184]: refactor test case

File: modules/core/02-client/abci.go (hunk 1 above)
Comment (seantking) [reply modules/core/02-client/abci.go:32]: Ah right, got it thanks. I'll update the godoc :+1:

File: modules/core/module.go [hunk 7]
```diff
@@ -164,7 +164,6 @@ func (AppModule) ConsensusVersion() uint64 { return 2 }
 
 // BeginBlock returns the begin blocker for the ibc module.
 func (am AppModule) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
-	ibcclient.BeginBlocker(ctx, am.keeper.ClientKeeper)
```
Comment (colin-axner) [inline modules/core/module.go:167]: add back

File: modules/core/02-client/types/genesis_test.go [hunk 8]
```diff
@@ -141,33 +134,13 @@ func (suite *TypesTestSuite) TestValidateGenesis() {
 			),
 			expPass: false,
 		},
-		{
-			name: "invalid client",
```
Comment (colin-axner) [inline modules/core/02-client/types/genesis_test.go:145]: why was this test case removed?

**Review by colin-axner — CHANGES_REQUESTED** [review]
Great work! Migration code is needed to remove any existing localhost from state otherwise removing all the `if clientID == localhost` checks could be problematic. It may even result in a panic if the clientID is ever referenced from store (we would no longer be able to unmarshal the client state)

File: modules/core/02-client/keeper/keeper_test.go (hunk 6 above)
Comment (seantking) [reply modules/core/02-client/keeper/keeper_test.go:184]: I thought the point here was that localhost is the invalid client type?

Comment (colin-axner) [reply modules/core/02-client/keeper/keeper_test.go:184]: yea but so is the solo machine. `ValidateSelfClient` must be called with a tendermint client

File: modules/core/02-client/types/genesis_test.go (hunk 8 above)
Comment (seantking) [reply modules/core/02-client/types/genesis_test.go:145]: I assumed the light client was invalid because it was a localhost client in this case. 

I can update the test case name to be more descriptive (to not confuse others) and put it back in.

File: modules/core/02-client/keeper/keeper_test.go (hunk 6 above)
Comment (seantking) [reply modules/core/02-client/keeper/keeper_test.go:184]: OK yeah makes sense.

File: modules/core/02-client/types/genesis_test.go (hunk 8 above)
Comment (colin-axner) [reply modules/core/02-client/types/genesis_test.go:145]: Hmm, that might be true, but maybe we can use it for `!ps.Params.AllowedClient` check

Comment (seantking) [reply modules/core/02-client/types/genesis_test.go:145]: Updated :handshake:

File: modules/core/02-client/abci_test.go [hunk 9]
```diff
@@ -30,65 +22,8 @@ func (suite *ClientTestSuite) SetupTest() {
...
-			client.BeginBlocker(suite.chainA.GetContext(), suite.chainA.App.GetIBCKeeper().ClientKeeper)
-		}, "BeginBlocker shouldn't panic")
-
-		localHostClient = suite.chainA.GetClientState(exported.Localhost)
-		suite.Require().Equal(prevHeight.Increment(), localHostClient.GetLatestHeight())
-		prevHeight = localHostClient.GetLatestHeight().(types.Height)
-	}
-}
-
-func (suite *ClientTestSuite) TestBeginBlockerConsensusState() {
```
Comment (colin-axner) [inline modules/core/02-client/abci_test.go:66]: add back test

File: modules/core/02-client/keeper/client.go [hunk 10]
```diff
@@ -98,11 +95,8 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, header exported.C
 	// Else the update was proof of misbehaviour and we must emit appropriate misbehaviour events.
 	if status := newClientState.Status(ctx, clientStore, k.cdc); status != exported.Frozen {
 		// if update is not misbehaviour then update the consensus state
-		// we don't set consensus state for localhost client
-		if header != nil && clientID != exported.Localhost {
+		if header != nil {
```
Comment (colin-axner) [inline modules/core/02-client/keeper/client.go:98]: remove if statement, header was only nil for localhost

**Review by colin-axner — COMMENTED** [review]
As a note, these changes are not correct until #1201 is addressed, so we should keep that in mind if we merge into 02-client-refactor

File: modules/core/02-client/keeper/client.go (hunk 10 above)
Comment (seantking) [reply modules/core/02-client/keeper/client.go]: Done

File: modules/core/02-client/abci_test.go (hunk 9 above)
Comment (seantking) [reply modules/core/02-client/abci_test.go]: Done :+1: good catch

**Review by colin-axner — APPROVED** [review]
ACK

**Review by damiannolan — APPROVED** [review]
